### DISCUSSION TITLE
electromagnet and crazyflie

### DISCUSSION THREAD

**Discussion by seunghyunnn** [bitcraze/discussions#115] [Q&A]
Hi, 

I was trying to implement my idea of using the electromagnet with the crazyflie to drop the paper clips in different spaces. 

To elaborate more on it, 
(1) I am aiming to use 2- 3 different electromagnets and attach a paper clip to each of them when the current is produced (when they have the magnetic characteristic). The electromagnets are each connected to the 12v battery and the transistor (so I will have a total of 2-3 batteries and 2-3 transistors) and the current will be provided by a controller (either controller of crazyflie, or controller attached to it) 
(2) Then I want my crazyflie to hover different places and drop each paper clip to each place by cutting off the current to each electromagnet using transistor.  

(Q) Is there a way to control multiple electromagnets like above using the built-in controller of crazyflie? (so basically, I need to control the 2-3 transistors)
If it is not possible, and therefore if I need an extra controller, can you recommend to me the way to use the extra controller and keep that weight under 15g? 

I am also open to any further suggestions on different ways other than using electromagnets/controller to drop the clips to different places with the crazyflie. 

Thanks a lot!

**Comment by tobbeanton** [comment]
You can use the expansion deck pins to output a 3V signal up to 20mA. It will probably not be enough for an electromagnet to hold a paper clip, you would have to investigate. If not, use a transistor (mosfet) as a switch that would supply the battery voltage (VBAT) to the coils and that would be activated using any of the free deck pins. The [simple deck API example](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/userguides/deck/) you could use as a base.

**Reply by seunghyunnn** [reply]
Thank you for the reply! I will test both of the suggestions that you have provided. 
Can I ask you some more questions? 
1. If I do not use the deck (like breakout deck) can I still use the pins in the crazyflie board to control the transistor by just connecting each circuit with transistor to the onboard pins(PB8, PB5, PB4, PB12 which are the pins for IO according to https://www.bitcraze.io/documentation/hardware/crazyflie_2_0/crazyflie_2_0-datasheet.pdf in the crazyflie? or will it be too complicated/impossible as we cannot mock/implement something like DECK_DRIVER() in the [simple deck API example](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/userguides/deck/) that you gave? 

2. I am currently using python code to fly the crazyflie based on https://github.com/bitcraze/crazyflie-lib-python . If I am to use the [simple deck API example](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/userguides/deck/)  how should I integrate this c code with my python code?  


3. Do you think using the electromagnet will disturb the crazyflie in any way? For example, since it has the magnetic characteristic, using electromagnets might disrupt some sensors in the crazyflie / since it is using the crazyflie's battery, it affects the propeller motor of crazyflie, etc


Thank you again:)

**Reply by tobbeanton** [reply]
1. You can still use the deck API even though you don't use a deck board. You just have to "force" the driver to run as described in the simple example.
2. You could create a parameter that would control the output. Then use the PARAM_ADD_WITH_CALLBACK macro to have a function called when the parameter is changed. In this function set the pin output to the parameter value. See [led.c](https://github.com/bitcraze/crazyflie-firmware/blob/0f36a866955074c7a267deec0d69399e8dcc11be/src/drivers/src/led.c#L131) as en example.
3. Shouldn't be a problem. Weight and possible power consumption are probably the biggest hurdles.
